### PR TITLE
[SPARK-32069][CORE][SQL] Improve error message on reading unexpected directory

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Improve error message on reading unexpected directory

### Why are the changes needed?
Improve error message on reading unexpected directory

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Ut

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
Ping @gengliangwang

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala [hunk 1]
```diff
@@ -232,6 +232,11 @@ class HadoopRDD[K, V](
         logWarning(s"${jobConf.get(FileInputFormat.INPUT_DIR)} doesn't exist and no" +
             s" partitions returned from this path.", e)
         Array.empty[Partition]
+      case e: IOException if e.getMessage.contains("Not a file") =>
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala:235]: Nit: if e.getMessage.startsWith("Not a file:")

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala [hunk 2]
```diff
@@ -232,6 +232,11 @@ class HadoopRDD[K, V](
...
+        val path = e.getMessage.split(":").map(_.trim).apply(2)
+        throw new IOException(s"Path: ${path} is a directory, it is not" +
+          s" allowed for `serde` reader when" +
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala:238]: The code path here is in core module. The reader may not be `serde` reader.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 3]
```diff
@@ -287,7 +287,11 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
               val msg = intercept[IOException] {
                 sql("SELECT * FROM tbl1").show()
               }.getMessage
-              assert(msg.contains("Not a file:"))
+              assert(msg.contains(
```
Comment (gengliangwang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala:290]: how about creating a method for checking the exception message?

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: > Nit: if e.getMessage.startsWith("Not a file:")

Done

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: > The code path here is in core module. The reader may not be `serde` reader.

How about current?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: > how about creating a method for checking the exception message?

Done

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
@dongjoon-hyun Seems jenkins down because of this pr? emmmmm

**Comment by HyukjinKwon** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala (hunk 2 above)
Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: How about:
```
s"Path: ${path} is a directory, which is not supported by the record reader when `mapreduce.input.fileinputformat.input.dir.recursive` is false." 
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala [hunk 4]
```diff
@@ -230,6 +230,13 @@ class HiveParquetSourceSuite extends ParquetPartitioningTest {
         withTempPath { path =>
           withTable("parq_tbl1", "parq_tbl2", "parq_tbl3",
             "tbl1", "tbl2", "tbl3", "tbl4", "tbl5", "tbl6") {
+
+            def errorMsg(path: String): String = {
```
Comment (gengliangwang) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala:234]: nit: `checkErrorMsg`

**Comment by gengliangwang** [comment]
LGTM otherwise. Thanks for doing this work.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
test this please

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: > How about:
> 
> ```
> s"Path: ${path} is a directory, which is not supported by the record reader when `mapreduce.input.fileinputformat.input.dir.recursive` is false." 
> ```

updated

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveParquetSourceSuite.scala]: Done

**Comment by srowen** [comment]
Seems OK. Slightly hacky to rely on the message text, and I guess you lose the underlying exception, but, both of those seem worth the upside.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please

**Review by dongjoon-hyun — COMMENTED** [review]
The failure looks relevant. Could you check it with `hadoop-2.7`?
```
- SPARK-25993 CREATE EXTERNAL TABLE with subdirectories *** FAILED ***
  Expected exception java.io.IOException to be thrown, but no exception was thrown (HiveParquetSourceSuite.scala:293)
```

**Comment by AngersZhuuuu** [comment]
> The failure looks relevant. Could you check it with `hadoop-2.7`?
> 
> ```
> - SPARK-25993 CREATE EXTERNAL TABLE with subdirectories *** FAILED ***
>   Expected exception java.io.IOException to be thrown, but no exception was thrown (HiveParquetSourceSuite.scala:293)
> ```

Yea,  checked a lot and run test with hadoop2.7 in local with success, and I have  checked the code of hadoop-2.7. and hadoop-3.2. seems no reason for this test failed..

**Comment by AngersZhuuuu** [comment]
> The failure looks relevant. Could you check it with `hadoop-2.7`?
> 
> ```
> - SPARK-25993 CREATE EXTERNAL TABLE with subdirectories *** FAILED ***
>   Expected exception java.io.IOException to be thrown, but no exception was thrown (HiveParquetSourceSuite.scala:293)
> ```

Found the reason, added UT `test("SPARK-32069: Improve error message on reading unexpected directory") ` 
run before `HiveParquetSourceSuite`, in UT `SPARK-32069` set conf  `sql("set mapreduce.input.fileinputformat.input.dir.recursive=true")` can impact  latter UT(maybe set tp system props? I will check more).
Seems this behavior is different between hadoop-2.7 and hadoop-3.2.

**Comment by dongjoon-hyun** [comment]
Ya. It bites us frequently. So, we need to pass during the review phase for the suspected PR.
> Seems this behavior is different between hadoop-2.7 and hadoop-3.2.

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
> **[Test build #129932 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129932/testReport)** for PR 30027 at commit [`d5947e1`](https://github.com/apache/spark/commit/d5947e1dea2ecb28bab7e6a5abb1eda80ff2b5ef).
> 
> * This patch **fails Spark unit tests**.
> * This patch merges cleanly.
> * This patch adds no public classes.

Failed caused by https://issues.apache.org/jira/browse/SPARK-33173

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
ping @dongjoon-hyun

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @AngersZhuuuu .
Merged to master.

**Pull request closed without merging**